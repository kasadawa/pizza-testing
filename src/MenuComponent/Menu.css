.resize{
    width: 300px;
    height: 200px;
    margin-left:10px;
}
.media{
    margin:20px; 
    border:1px solid; 
    border-radius: 13px; 
}
i{
    padding-left: 10px;
    padding-right: 10px;
    
}