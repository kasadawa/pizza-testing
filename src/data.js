export var pizzas = [{
    name:'Margarita',
    image: 'margarita.jpg',
    quantity: 1, 
    descr:'tomato, and other ingredience',
    style:'normal',
    options: [],
},{
    name:'Devil pizza',
    image: 'deliv.jpg',
    quantity: 1,
    descr:'ham,tomato, and other ingredience',
    style:'normal',
    options: []
},{
    name:'Calcone',
    descr:'mushrooms,ham,tomato, and other ingredience',
    quantity: 1,
    image: 'calcone.jpg',
    style:'normal',
    options: []
}];

export var prices = new Map();
prices.set('Margarita',12) 
prices.set('Devil pizza',14.4) 
prices.set('Calcone',15)
prices.set('italian', 2) 
prices.set('exstra cheese', 1.3) 
prices.set('exstra tomato', 3)
