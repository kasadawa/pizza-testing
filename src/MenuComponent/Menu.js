import  React from 'react';
import { pizzas,prices } from '../data';
import margarita  from '../img/margarita.jpg';
import devil  from '../img/devil.jpg';
import calcone  from '../img/calcone.jpg';

import './Menu.css';


// var orderPizzaObj = {
//     extraTomato: Bool, 
//     extraCheese: Bool,
//     pizzaName:String, 
//     totalPrice: Number,
//     type:String,
//     quantity: Number, 
// }



// get Array and clone its objects 
// deep cloning
function cloneArr(arr){
    return arr.map((val)=>{
        if(typeof val === 'object'){
            return cloneObj(val);
        }
    })
}

function cloneObj(obj){
    var copy = {} ; 
    for(let key in obj){
        if(obj.hasOwnProperty(key)){
            if(typeof obj[key] == 'object' && !Array.isArray(obj[key])){
                copy[key] = cloneObj(obj[key]);
            }else if( Object.prototype.toString.call(obj[key]) == '[object Array]'){
                copy[key] =[];
            }else{
                copy[key] = obj[key];
            }
        }
    }
    return copy; 
}

export default class Menu extends React.Component {

    constructor(props){
        super(props);
        [pizzas[0].image, pizzas[1].image,pizzas[2].image] = [margarita,devil,calcone];

        this.state =  cloneArr(pizzas); // this can be reworked
        


        // bindings
        this.changeIngredient = this.changeIngredient.bind(this);
       
    }

    changeQuantity(key, type){ 
        if(!isNaN(this.state[key].quantity) && !(this.state[key].quantity === 1 && type ==='dec') ){
           
            this.setState((state)=>{
                state[key].quantity =  type === 'inc' ? state[key].quantity+ 1 : state[key].quantity- 1; 
                return state;
            })
        }
    }

    changeStyle(key,e){
        this.state[key].style = e.target.value;
        this.setState(this.state);
    }

    changeIngredient(key,e){
        console.log('state',this.state)
        if(e.target.checked){
            this.state[key].options.push(e.target.value);
        }else{
            let i = this.state[key].options.indexOf(e.target.value);
            this.state[key].options.splice(i,1);
        }
        this.setState(this.state);
    }
    
    addToBasket(key){
        var orderNum = window.sessionStorage.getItem('ordernum'); 
        // +null = 0 :O 
        if(isNaN(+orderNum) || orderNum == null ){ // if someone changed the sessionStorage or its null
            window.sessionStorage.setItem('ordernum',1);
            orderNum = 1 ;
        }else{
            orderNum++;
        }
        window.sessionStorage.setItem('order' + orderNum ,JSON.stringify(this.state[key]));
        window.sessionStorage.setItem('ordernum',orderNum);
        this.setState(cloneArr(pizzas));
    }

    orderPizza(){

    }

  render() {
    return (
      <div>
          { pizzas.map((pizza,key)=>{
              
              return <div key={key} className="media">
              <img src={this.state[key].image} className="mr-3 resize" />
              <div className="media-body">
                <h5 className="mt-0">{pizza.name}</h5>
                Ingredients: {pizza.descr}

                <div className="col-md-4">
                <p>Quantity <span>{this.state[key].quantity}</span><i onClick={this.changeQuantity.bind(this,key,'inc')} className="fas fa-plus"></i><i onClick={this.changeQuantity.bind(this,key,'dec')} className="fas fa-minus"></i></p>
                </div>
                <div className='row'>
                <div className="col-md-4">
                    <div className="input-group mb-3">
                    <div className="input-group-prepend">
                        <label className="input-group-text" htmlFor="inputGroupSelect01">Pizza Style</label>
                    </div>
                    <select onChange={(e)=>this.changeStyle(key,e)} className="custom-select" id="inputGroupSelect01">
                        <option value="clasic" selected>clasic</option>
                        <option value="italian">italian</option>
                    </select>
                    </div>
                </div>
                <div className="col-md-4"></div>
                <div className="col-md-4">
                    <button onClick={this.addToBasket.bind(this,key)} className="btn btn-success" >Add to basket</button>
                </div>
                </div>
                <div className="col-md-4">

                <div className="input-group mb-3">
           

                    <div className="form-check">
                        <input onChange={(e)=>this.changeIngredient(key,e)} className="form-check-input" type="checkbox" value="cheese" id={"cheese"+ key }/>
                        <label className="form-check-label"  htmlFor={ "cheese" + key}>
                            Extra Cheese
                        </label>
                        <br></br>
                        <input onChange={(e)=>this.changeIngredient(key,e)} className="form-check-input" type="checkbox" value="tomato" id={"tomato"+ key} />
                        <label className="form-check-label" htmlFor={"tomato" + key}>
                            Extra Tomato
                        </label>
                    </div>

                    </div>
                  
                </div>

              </div>



            </div>;
          })

          }


      </div>
    );
  }
}