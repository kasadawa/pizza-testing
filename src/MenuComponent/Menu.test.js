import React from 'react';
import renderer from 'react-test-renderer';
import Enzyme, {shallow,render} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';


Enzyme.configure({adapter: new Adapter()});

import Menu from './Menu';
import { pizzas,prices } from '../data';

test('Order changes the class when hovered', () => {
  const component = renderer.create(
    <Menu></Menu>,
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();

});

describe('Testing Menu component adding and removing pizzas',()=>{
  it('Add one more pizza to every pizza type', () => {
  
    var context = {};
    const wrapper = shallow(<Menu/>);
    wrapper.setState({}); 
  
  
    expect(wrapper.find('.mitko')).toHaveLength(1);
    expect(wrapper.find('.media')).toHaveLength(3);
  
  
    // check all pizza objects 
    for(let i = 0 ; i < Object.keys(wrapper.state()).length; i++ ){
      expect(wrapper.state()[i].quantity).toBe(1);
     
      // expect(wrapper.state()[i].quantity).toBe(2);
  
    }
  
  
    wrapper.find('i.fas.fa-plus').forEach((node,index)=>{
      node.simulate('click');
      expect(wrapper.state()[index].quantity).toBe(2);
    })

  });

});

 

