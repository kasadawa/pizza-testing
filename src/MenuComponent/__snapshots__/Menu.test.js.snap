// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Order changes the class when hovered 1`] = `
<div
  className="mitko"
>
  <div
    className="media"
  >
    <img
      className="mr-3 resize"
      src="margarita.jpg"
    />
    <div
      className="media-body"
    >
      <h5
        className="mt-0"
      >
        Margarita
      </h5>
      Ingredients: 
      tomato, and other ingredience
      <div
        className="col-md-4"
      >
        <p>
          Quantity 
          <span>
            1
          </span>
          <i
            className="fas fa-plus"
            onClick={[Function]}
          />
          <i
            className="fas fa-minus"
            onClick={[Function]}
          />
        </p>
      </div>
      <div
        className="row"
      >
        <div
          className="col-md-4"
        >
          <div
            className="input-group mb-3"
          >
            <div
              className="input-group-prepend"
            >
              <label
                className="input-group-text"
                htmlFor="inputGroupSelect01"
              >
                Pizza Style
              </label>
            </div>
            <select
              className="custom-select"
              id="inputGroupSelect01"
              onChange={[Function]}
            >
              <option
                selected={true}
                value="clasic"
              >
                clasic
              </option>
              <option
                value="italian"
              >
                italian
              </option>
            </select>
          </div>
        </div>
        <div
          className="col-md-4"
        />
        <div
          className="col-md-4"
        >
          <button
            className="btn btn-success"
            onClick={[Function]}
          >
            Add to basket
          </button>
        </div>
      </div>
      <div
        className="col-md-4"
      >
        <div
          className="input-group mb-3"
        >
          <div
            className="form-check"
          >
            <input
              className="form-check-input"
              id="cheese0"
              onChange={[Function]}
              type="checkbox"
              value="cheese"
            />
            <label
              className="form-check-label"
              htmlFor="cheese0"
            >
              Extra Cheese
            </label>
            <br />
            <input
              className="form-check-input"
              id="tomato0"
              onChange={[Function]}
              type="checkbox"
              value="tomato"
            />
            <label
              className="form-check-label"
              htmlFor="tomato0"
            >
              Extra Tomato
            </label>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="media"
  >
    <img
      className="mr-3 resize"
      src="devil.jpg"
    />
    <div
      className="media-body"
    >
      <h5
        className="mt-0"
      >
        Devil pizza
      </h5>
      Ingredients: 
      ham,tomato, and other ingredience
      <div
        className="col-md-4"
      >
        <p>
          Quantity 
          <span>
            1
          </span>
          <i
            className="fas fa-plus"
            onClick={[Function]}
          />
          <i
            className="fas fa-minus"
            onClick={[Function]}
          />
        </p>
      </div>
      <div
        className="row"
      >
        <div
          className="col-md-4"
        >
          <div
            className="input-group mb-3"
          >
            <div
              className="input-group-prepend"
            >
              <label
                className="input-group-text"
                htmlFor="inputGroupSelect01"
              >
                Pizza Style
              </label>
            </div>
            <select
              className="custom-select"
              id="inputGroupSelect01"
              onChange={[Function]}
            >
              <option
                selected={true}
                value="clasic"
              >
                clasic
              </option>
              <option
                value="italian"
              >
                italian
              </option>
            </select>
          </div>
        </div>
        <div
          className="col-md-4"
        />
        <div
          className="col-md-4"
        >
          <button
            className="btn btn-success"
            onClick={[Function]}
          >
            Add to basket
          </button>
        </div>
      </div>
      <div
        className="col-md-4"
      >
        <div
          className="input-group mb-3"
        >
          <div
            className="form-check"
          >
            <input
              className="form-check-input"
              id="cheese1"
              onChange={[Function]}
              type="checkbox"
              value="cheese"
            />
            <label
              className="form-check-label"
              htmlFor="cheese1"
            >
              Extra Cheese
            </label>
            <br />
            <input
              className="form-check-input"
              id="tomato1"
              onChange={[Function]}
              type="checkbox"
              value="tomato"
            />
            <label
              className="form-check-label"
              htmlFor="tomato1"
            >
              Extra Tomato
            </label>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="media"
  >
    <img
      className="mr-3 resize"
      src="calcone.jpg"
    />
    <div
      className="media-body"
    >
      <h5
        className="mt-0"
      >
        Calcone
      </h5>
      Ingredients: 
      mushrooms,ham,tomato, and other ingredience
      <div
        className="col-md-4"
      >
        <p>
          Quantity 
          <span>
            1
          </span>
          <i
            className="fas fa-plus"
            onClick={[Function]}
          />
          <i
            className="fas fa-minus"
            onClick={[Function]}
          />
        </p>
      </div>
      <div
        className="row"
      >
        <div
          className="col-md-4"
        >
          <div
            className="input-group mb-3"
          >
            <div
              className="input-group-prepend"
            >
              <label
                className="input-group-text"
                htmlFor="inputGroupSelect01"
              >
                Pizza Style
              </label>
            </div>
            <select
              className="custom-select"
              id="inputGroupSelect01"
              onChange={[Function]}
            >
              <option
                selected={true}
                value="clasic"
              >
                clasic
              </option>
              <option
                value="italian"
              >
                italian
              </option>
            </select>
          </div>
        </div>
        <div
          className="col-md-4"
        />
        <div
          className="col-md-4"
        >
          <button
            className="btn btn-success"
            onClick={[Function]}
          >
            Add to basket
          </button>
        </div>
      </div>
      <div
        className="col-md-4"
      >
        <div
          className="input-group mb-3"
        >
          <div
            className="form-check"
          >
            <input
              className="form-check-input"
              id="cheese2"
              onChange={[Function]}
              type="checkbox"
              value="cheese"
            />
            <label
              className="form-check-label"
              htmlFor="cheese2"
            >
              Extra Cheese
            </label>
            <br />
            <input
              className="form-check-input"
              id="tomato2"
              onChange={[Function]}
              type="checkbox"
              value="tomato"
            />
            <label
              className="form-check-label"
              htmlFor="tomato2"
            >
              Extra Tomato
            </label>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
